inc/az_ulib_ustream_forward_base.h

```diff
@@ -0,0 +1,374 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license.
+// See LICENSE file in the project root for full license information.
+
+#ifndef AZ_ULIB_ustream_forward_BASE_H
+#define AZ_ULIB_ustream_forward_BASE_H
+
+#include "az_ulib_config.h"
+#include "az_ulib_pal_os.h"
+#include "az_ulib_result.h"
+#include "azure/core/az_span.h"
+#include "azure/core/az_context.h"
+
+#ifdef __cplusplus
+#include <cstddef>
+#include <cstdint>
+#else
+#include <stddef.h>
+#include <stdint.h>
+#endif /* __cplusplus */
+
+#include "azure/core/_az_cfg_prefix.h"
+
+/**
+ * @brief   Define offset_t with the same size as size_t.
+ */
+typedef size_t offset_t;
+
+/**
+ * @brief   Forward declaration of az_ulib_ustream_forward. See #az_ulib_ustream_forward_tag for struct members.
+ */
+typedef struct az_ulib_ustream_forward_tag az_ulib_ustream_forward;
+
+/**
+ * @brief Signature of the function to be invoked by the `ustream_forward->flush` operation when the 
+ *        `const az_span* const` buffer has been created.
+ * 
+ * @param[in]   buffer                  The `const az_span* const` buffer to be handled by the 
+ *                                      implementation of this callback.
+ * @param[in]   push_callback_context   The #az_context* contract held between the owner of this
+ *                                      callback and the `ustream_forward->flush` operation.
+ */
+typedef void (*az_ulib_flush_callback)(const az_span* const buffer, az_context* push_callback_context);
+
+/**
+ * @brief   vTable with the ustream_forward APIs.
+ *
+ *  Any module that exposes the ustream_forward shall implement the functions on this vTable.
+ *
+ *  Any code that will use an exposed ustream_forward shall call the APIs using the `az_ulib_ustream_forward_...`
+ *      inline functions.
+ */
+typedef struct az_ulib_ustream_forward_interface_tag
+{
+  /** Concrete `flush` implementation. */
+  az_result (*flush)(
+      az_ulib_ustream_forward* ustream_forward_instance, 
+      az_ulib_flush_callback push_callback, 
+      az_context* push_callback_context);
+
+  /** Concrete `read` implementation. */
+  az_result (*read)(
+      az_ulib_ustream_forward* ustream_forward_instance,
+      uint8_t* const buffer,
+      size_t buffer_length,
```

Comment:
Lets unify these APIs to use `az_span` or `uint8_t*` with `size_t`.

Reply:
they are currently unified using `uint8_t*` and `size_t` as they are in ustream. are you suggesting we move to az_span?